### DISCUSSION TITLE
In what circumstances did qsim return incorrect results in 22.11?

### DISCUSSION THREAD

**Discussion by eliottrosenberg** [NVIDIA/cuQuantum#58] [Q&A] [answered]
I saw that the [release notes](https://docs.nvidia.com/cuda/cuquantum/appliance/release_notes.html) for cuQuantum Appliance 23.03 say "Fixed an issue that qsim can return incorrect results when multiple GPUs are used for simulation."

Under what circumstances were the incorrect results seen? I used cuQuantum Appliance 22.11 with qsim, using multiple GPUs, to take simulation data for an experiment and want to make sure that I don't need to go back and check the quality of the data.

Thanks!

**Comment by leofang** [comment]
Hi @eliottrosenberg, it refers to a delicate synchronization bug that would not be easily triggered unless a frequent data exchange between GPUs happens, which is problem dependent.

The bug was fixed in part with [cuStateVec v1.3.0](https://docs.nvidia.com/cuda/cuquantum/custatevec/release_notes.html#custatevec-v1-3-0) (included as part of the 23.03 release). The relevant entry is:
> Fix for an issue of [custatevecMultiDeviceSwapIndexBits()](https://docs.nvidia.com/cuda/cuquantum/custatevec/api/functions.html#_CPPv434custatevecMultiDeviceSwapIndexBitsP18custatevecHandle_tK8uint32_tPPvK14cudaDataType_tK8uint32_tK8uint32_tPK4int2K8uint32_tPK7int32_tPK7int32_tK8uint32_tK29custatevecDeviceNetworkType_t) that CUDA calls are not correctly ordered on streams allocated on multiple GPUs.

**Reply by eliottrosenberg** [reply]
Thanks, @leofang! In my case, I was running circuits on 36 qubits using 8 80-GB NVIDIA A100 GPUs and sampling bitstrings at the end. Do you think I need to go back and re-collect this data using the latest version of cuQuantum Appliance?

**Reply by mtjrider** [reply]
By "problem," @leofang meant the circuit itself. 

I suggest you rerun your workload on a smaller subset (fewer samples, etc.) to confirm qualitative behavior is unchanged. If the results are qualitatively similar, you most likely did not trigger the issue.

**Reply by leofang** [reply]
It certainly would not hurt to exercise caution and rerun the calculation if resource permits. As I said it's problem (circuit) dependent, so it's hard to say if you're impacted or not. It'd be great if you could share with your findings with us (and the audience here) should you rerun 🙂

**Reply by eliottrosenberg** [reply]
Thanks for answering :)

First, we applied X gates to some of the qubits (on average half of them) to prepare a bitstring state. Then, we applied 18 cycles of fSim gates, where an fSim cycle is

```python
def fsim_cycle(theta, phi, n):
    
    qc = cirq.Circuit( cirq.FSimGate(theta, phi).on(cirq.LineQubit(q), cirq.LineQubit(q+1)) for q in range(0,n-1,2) ) +  cirq.Circuit( cirq.FSimGate(theta, phi).on(cirq.LineQubit(q), cirq.LineQubit(q+1)) for q in range(1,n-1,2) )
    
    return qc
```

That is, one cycle looks like this, except we use 36 qubits instead of 8.
```
0: ───FSim(θ, φ)────────────────
      │
1: ───FSim(θ, φ)───FSim(θ, φ)───
                   │
2: ───FSim(θ, φ)───FSim(θ, φ)───
      │
3: ───FSim(θ, φ)───FSim(θ, φ)───
                   │
4: ───FSim(θ, φ)───FSim(θ, φ)───
      │
5: ───FSim(θ, φ)───FSim(θ, φ)───
                   │
6: ───FSim(θ, φ)───FSim(θ, φ)───
      │
7: ───FSim(θ, φ)────────────────
```

Then we sample bitstrings at the end. $\theta$ and $\phi$ are each uniform across all of the gates.

**Reply by eliottrosenberg** [reply]
The results matched what we expected from studying earlier cycles without using multiple GPUs.

**Reply by eliottrosenberg** [reply]
Do you happen to know whether this is a circuit that would be affected?

**Reply by eliottrosenberg** [reply]
I just re-ran some of my 36-qubit circuits in cuQuantum Appliance v23.03, and the results are consistent with what I found using v22.11.

**Reply by mtjrider** [reply, marked as answer]
Sorry for not responding sooner and thanks for rerunning your workload!

I would have been surprised if your results were affected given the circuit's local structure. 

This is an excellent datapoint for us. Very much appreciated.

**Reply by eliottrosenberg** [reply]
@leofang @mtjrider Thanks for your help with this! The project that I was referring to is now available as an arXiv preprint: https://arxiv.org/abs/2306.09333. I'm happy to answer questions about this work or present it to your group if you're interested.
